### PR TITLE
Add best practices for bottom up approach

### PR DESCRIPTION
**Pull request by shimks**
- connected to https://github.com/strongloop/loopback-next/issues/735
- initial docs proposal for adding a 'best practices' section for bottom up approach
- need link to the `getting-started-example` once that's added into the docs

### CODE REVIEW

File: pages/en/lb4/Best-practices-with-Loopback-4.md [hunk 1]
```diff
@@ -12,8 +12,10 @@ LoopBack 4 is more than just a framework: It’s an ecosystem that encourages de
...
-2. [**Testing the API**](./Testing-the-API.html): This section describes the process of writing smoke test for your API and its spec.
-3. [**Defining your testing strategy**](./Defining-your-testing-strategy.html): This section discusses the advantages and the process of building a strong testing suite.
-4. [**Implementing features**](./Implementing-features.html): This section demonstrates how the tests for each feature of your application should be written, and how to write the logic to make these tests pass. In the example, the tests for the controller, model, repository, data source, and sequence are written and then implemented.
-5. [**Preparing the API for consumption**](./Preparing-the-API-for-consumption.html): This section shows how the endpoints can be physically tested using the Swagger UI.
+1. **Defining the API**: There are two possible approaches to take in this section
+  - [**Defining the API using bottom-up appraoch**](./Defining-the-API-using-bottom-up-approach.html): This section guides you through setting up a skeleton of your application so that a full API can be automatically generated.
```
Comment (kjdelisle) [inline pages/en/lb4/Best-practices-with-Loopback-4.md]: - Typo in `approach`
- Not sure what you mean by `so that a full API can be automatically generated`; the bottom-up approach is one where you're building the controllers, functions, routes and so on that make up the bulk of your API. The portion that's generated is the OpenAPI/Swagger specification that _describes_ it.

File: pages/en/lb4/Best-practices-with-Loopback-4.md [hunk 2]
```diff
@@ -12,8 +12,10 @@ LoopBack 4 is more than just a framework: It’s an ecosystem that encourages de
...
+  - [**Defining the API using top-down approach**](./Defining-the-API-using-top-down-approach.html): This section guides you through constructing your API first before any internal logic is added.
```
Comment (kjdelisle) [inline pages/en/lb4/Best-practices-with-Loopback-4.md]: Please follow the convention established by @b-admike 's PRs and mark this with a warning or indicator that says this approach is not yet functional/viable.

File: pages/en/lb4/Defining-the-API-using-bottom-up-approach.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  - `@property()`
+- [Routes](Routes.html)
+  - `@operation()`
+  - `@param()`
+
+### Define your models
+
+Your models will be used to provide schemas to validate against the data
+your API will intercept and consequently are likely to be referenced as
+argument types in your controllers, so they should be the first to be defined.
```
Comment (kjdelisle) [inline pages/en/lb4/Defining-the-API-using-bottom-up-approach.md]: - Your models are not necessarily going to be used to validate anything. I can easily write my controller functions to accept objects that are neither strictly nor logically equivalent to my model objects.
- The _value_ of using models is that they give you a common, workable definition for:
  - representing inbound and outbound data at the API layer
  - handling data and relationships at the datasource layer
- _Intercept_ carries a very particular connotation; that the intended recipient is not the first to receive or act upon the incoming payload. In our case, we're not intercepting the request, we're handling it.

File: pages/en/lb4/Defining-the-API-using-bottom-up-approach.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+export class Todo {
+  id?: number;
+  title: string;
+  desc?: string;
+  isComplete: boolean;
+}
+```
+
+This model is just a TypeScript class without any decorators in its current
+form, so features such as automatic schema generation are not available to it
```
Comment (kjdelisle) [inline pages/en/lb4/Defining-the-API-using-bottom-up-approach.md]: This is just filler. It's more than adequate to say that this is just a plain TypeScript class.

File: pages/en/lb4/Defining-the-API-using-bottom-up-approach.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+yet.
+
+Now decorate the class with `@model` and `@property`.
```
Comment (kjdelisle) [inline pages/en/lb4/Defining-the-API-using-bottom-up-approach.md]: Explain what these decorators accomplish in the same sentence/paragraph here.
```
We can use the @model and @property decorators to create _metadata_ that
represents the model; a model definition. LoopBack and LoopBack extensions
can use this model definition for a wide variety of uses, such as:
- generating OpenAPI schema for your APIs
- validating instances of the models during the request/response lifecycle
- automatically inferring relationships between models during datasource
   operations

To apply these decorators to your model, you simply prefix the class definition
with the @model decorator, and prefix each property with the @property decorator:
```

**Comment by virkt25** [comment]
Haven't reviewed the entire PR but a high level question ... 

Do people normally understand what is meant by `top-down` and `bottom-up` approach? It personally took me a while to remember which approach is which and even now at times I have to think about it. I think we need better names (but don't know any off the top of my head).

**Comment by virkt25** [comment]
Whoops. Accidentally hit `Close and comment`.

File: pages/en/lb4/Defining-the-API-using-bottom-up-approach.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+  })
+  title: string;
+  @property() desc?: string;
+  @property() isComplete: boolean;
+}
+```
+
+Once the model has been decorated with these decorators,
+its metadata is made available to the app and the model is ready to have its
+corresponding OpenAPI schema generated.
```
Comment (kjdelisle) [inline pages/en/lb4/Defining-the-API-using-bottom-up-approach.md]: Remove this sentence. Technically speaking:
- you've already covered this (if you implement my feedback)
- the decorators are built up at compile time into JS, and that JS runs as a part of the parsing of classes and properties, which is before the application has actually begun to run through boot phases, or startup.

File: pages/en/lb4/Defining-the-API-using-bottom-up-approach.md [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+### Define your routes
+
+{% include note.html content="
+  `TodoController` from [tutorial](TUTORIAL LINK) is used for
+  demonstration here.
+" %}
+
+Once your models are defined, create a controller to host your routes
+for each categories of your API.
+
```
Comment (kjdelisle) [inline pages/en/lb4/Defining-the-API-using-bottom-up-approach.md]: There should be a tag here to indicate the name of this example file.
`{% include code-caption.html content="src/todo.controller.ts" %}`

File: pages/en/lb4/Defining-the-API-using-bottom-up-approach.md [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  // ...
+}
+```
+
+The controller's routes in their current state has no information on which
+API endpoints they belong to. Add them in by using `@operation` and `@param`
+decorators.
+
```
Comment (kjdelisle) [inline pages/en/lb4/Defining-the-API-using-bottom-up-approach.md]: There should be a tag here to indicate the name of this example file.
`{% include code-caption.html content="src/todo.controller.ts" %}`

File: pages/en/lb4/Defining-the-API-using-bottom-up-approach.md [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+{% include note.html content="
+  `TodoController` from [tutorial](TUTORIAL LINK) is used for
+  demonstration here.
+" %}
+
+Once your models are defined, create a controller to host your routes
+for each categories of your API.
+
+```ts
+class TodoController {
```
Comment (kjdelisle) [inline pages/en/lb4/Defining-the-API-using-bottom-up-approach.md]: `export`

File: pages/en/lb4/Defining-the-API-using-bottom-up-approach.md [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+  // ...
+}
+```
+
+The controller's routes in their current state has no information on which
+API endpoints they belong to. Add them in by using `@operation` and `@param`
+decorators.
+
+```ts
+class TodoController {
```
Comment (kjdelisle) [inline pages/en/lb4/Defining-the-API-using-bottom-up-approach.md]: `export`

File: pages/en/lb4/Defining-the-API-using-bottom-up-approach.md [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+  // ...
+}
+```
+
+Once your routes have been decorated, your application is ready to serve
+its API; when an instance of `RestServer` is run, an OpenAPI specification
+representing your application's API is built. The spec is generated
+entirely from the decorated elements' metadata, which in turn provides
+routing logic for your API when your application is running.
```
Comment (kjdelisle) [inline pages/en/lb4/Defining-the-API-using-bottom-up-approach.md]: ✅

**Comment by shimks** [comment]
I think it's ok to keep the names, but I agree on the fact that those names need clarification on what we mean by them.

React seems to be using the terms (https://reactjs.org/docs/thinking-in-react.html#step-2-build-a-static-version-in-react), but not all that often. I'd like to hear other people's thoughts on this

File: pages/en/lb4/Testing-the-API.md [hunk 12]
```diff
@@ -9,7 +9,7 @@ summary:
 ---
 
 {% include previous.html content="
-This article continues off from [Defining and validating the API](./Defining-and-validating-the-API.html).
+This article continues off from [Defining the API using top-down approach](./Defining-the-API-using-top-down-approach.html).
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-the-API.md]: `continues from`

**Review by kjdelisle — CHANGES_REQUESTED** [review]
Mostly good, just a few typos and some areas of confusion that need tidying up.

File: pages/en/lb4/Best-practices-with-Loopback-4.md (hunk 2 above)
Comment (b-admike) [reply pages/en/lb4/Best-practices-with-Loopback-4.md]: I think the warning on the page is sufficient, since they will click the links and be redirected there. But if we can do it here as well, then I'm fine with it.

File: pages/en/lb4/Defining-the-API-using-bottom-up-approach.md [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+
+### Define your routes
+
+{% include note.html content="
+  `TodoController` from [tutorial](TUTORIAL LINK) is used for
+  demonstration here.
+" %}
+
+Once your models are defined, create a controller to host your routes
+for each categories of your API.
```
Comment (b-admike) [inline pages/en/lb4/Defining-the-API-using-bottom-up-approach.md]: I'd say  `for each [Paths object](https://swagger.io/specification/#pathsObject) of your API specification` here.

File: pages/en/lb4/Defining-the-API-using-bottom-up-approach.md [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+
+The controller's routes in their current state has no information on which
+API endpoints they belong to. Add them in by using `@operation` and `@param`
+decorators.
+
+```ts
+class TodoController {
+  constructor() {}
+
+  @post('/todo') // sugar for @operation('post', '/todo');
```
Comment (b-admike) [inline pages/en/lb4/Defining-the-API-using-bottom-up-approach.md]: not sure if people would know what `sugar` means, so I suggest you rephrase as `same as @operation('post', '/todo');`

File: pages/en/lb4/Defining-the-API-using-bottom-up-approach.md [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+Once your routes have been decorated, your application is ready to serve
+its API; when an instance of `RestServer` is run, an OpenAPI specification
+representing your application's API is built. The spec is generated
+entirely from the decorated elements' metadata, which in turn provides
+routing logic for your API when your application is running.
+
+### Reviewing your API specification
+
+To review your complete API specification, run your application with the
+decorated controllers registered. Once it is running, visit `/swagger.json`
```
Comment (b-admike) [inline pages/en/lb4/Defining-the-API-using-bottom-up-approach.md]: They can also go to `openapi.json` to get OpenAPI 3.0 spec and `.yaml` for both to get them in yaml flavour.

File: pages/en/lb4/Defining-the-API-using-bottom-up-approach.md [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+summary:
+---
+
+## Define the API from bottom to top
+
+### Start with LoopBack artfiacts
+
+Sometimes, it's hard to know what your API is going to look like without
+writing out your application first. It also might be difficult to adjust your
+API as plans change or as features need to be reworked.
```
Comment (raymondfeng) [inline pages/en/lb4/Defining-the-API-using-bottom-up-approach.md]: IMO, the reasoning for bottom-up is misleading here. Please rewrite it to align with some ideas below:

1. Designing APIs in vacuum is not only difficult but also not wise if you don't have a good understanding of what your existing system can offer and what your clients need.

2. Bottom-up approach allows developers to capture existing business domain models and corresponding operations and reflect them for external consumption as APIs. Such information can be defined as metadata with code, JSON/YAML documents, or decorations.

3. Bottom-up approach better aligns APIs with implementations from the beginning and evolve them all together as API consumers and providers reach agreements incrementally.

4. Bottom-up approach makes it possible to have APIs up and running for play and polish at very early stages, which is critical to the success of APIs as stakeholders can see the big picture collaboratively before it's too late.

File: pages/en/lb4/Defining-the-API-using-bottom-up-approach.md (hunk 15 above)
Comment (shimks) [reply pages/en/lb4/Defining-the-API-using-bottom-up-approach.md]: I want to leave it at just `swagger` for now since we don't support OpenAPI v3 yet, but I'll add in short sentence saying that yaml flavour is also available. I'll leave it up to Janny to fix the `swagger` to `openapi` related doc switch-overs when she's doing her docs.

File: pages/en/lb4/Defining-the-API-using-bottom-up-approach.md [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+
+- do not have a complete understanding of what your existing tools can offer.
+- want to capture already existing domain models so that they can be reflected
+as APIs for external consumption.
+- need to grow and change your API from the initial implementation
+- want to set up and run an API from an early stage of the production to
+easily envision the big picture of the end product.
+
+There are various tools available to LoopBack which allows this bottom-up
+approach of building your application simple through the usages of
```
Comment (b-admike) [inline pages/en/lb4/Defining-the-API-using-bottom-up-approach.md]: `simple` -> `simply`

**Comment by b-admike** [comment]
In terms of what to name the sections, how about `code first application design` vs `API first application design` or something along those lines?

**Review by b-admike — APPROVED** [review]

**Comment by bajtos** [comment]
The new content in Defining-the-API-using-code-first-approach.md does not say anything about testing your new application. Originally, Thinking in LoopBack was heavily focused on showing people how to write application in test-first/test-driven style. It would be great to preserve this spirit for code-first approach too.

**Comment by shimks** [comment]
@bajtos Could you be more specific as to what you're envisioning? The code in `Defining-the-API-using-code-first-approach.md` contains no logic, so do you mean that I should show a skeleton of the endpoints that I need to test out for the controller?

File: _data/sidebars/lb4_sidebar.yml [hunk 18]
```diff
@@ -103,13 +103,18 @@ children:
   output: 'web, pdf'
   children:
 
-  - title: 'Defining and validating the API'
-    url: Defining-and-validating-the-API.html
+  - title: 'Defining the API using code-first approach'
```
Comment (virkt25) [inline _data/sidebars/lb4_sidebar.yml:114]: `Defining API's using code-first approach` might be better. `Defining the API ...` seems odd.

File: _data/sidebars/lb4_sidebar.yml [hunk 19]
```diff
@@ -103,13 +103,18 @@ children:
...
+    url: Defining-the-API-using-code-first-approach.html
     output: 'web, pdf'
 
-  - title: 'Testing the API'
-    url: Testing-the-API.html
+  - title: 'Defining the API using API-first approach'
```
Comment (virkt25) [inline _data/sidebars/lb4_sidebar.yml]: Ditto

File: _data/sidebars/lb4_sidebar.yml [hunk 20]
```diff
@@ -103,13 +103,18 @@ children:
...
+    url: Defining-the-API-using-API-first-approach.html
     output: 'web, pdf'
+    children:
+
+    - title: 'Testing the API'
```
Comment (virkt25) [inline _data/sidebars/lb4_sidebar.yml]: `Testing API's` ... ?

File: pages/en/lb4/Best-practices-with-Loopback-4.md [hunk 21]
```diff
@@ -8,12 +8,21 @@ permalink: /doc/en/lb4/Best-practices-with-Loopback-4.html
...
-5. [**Preparing the API for consumption**](./Preparing-the-API-for-consumption.html): This section shows how the endpoints can be physically tested using the Swagger UI.
+1. **Defining the API**: There are two possible approaches to take in this section
+   - [**Defining the API using code-first approach**](./Defining-the-API-using-code-first-approach.html): This section guides you through setting up a skeleton of your application so that its full OpenAPI specification can be automatically generated.
+   - [**Defining the API using API-first approach**](./Defining-the-API-using-API-first-approach.html): This section guides you through constructing your API first before any internal logic is added. __*Not fully supported*__
+     - [**Testing the API**](./Testing-the-API.html): This section describes the process of writing smoke test for your API and its spec. __*Not fully supported*__
```
Comment (virkt25) [inline pages/en/lb4/Best-practices-with-Loopback-4.md]: Shouldn't this be under `Defining your testing strategy`?

**Comment by bajtos** [comment]
> Could you be more specific as to what you're envisioning? The code in Defining-the-API-using-code-first-approach.md contains no logic, so do you mean that I should show a skeleton of the endpoints that I need to test out for the controller?

Right now, your narrative is "Define your models", "Define your routes", "Reviewing your API specification". That's not how test-first people work! Take a look at [Incrementally implement features](https://github.com/strongloop/loopback.io/blob/7ccfb55463b36f9177a7960926f933ae47f00986/pages/en/lb4/Implementing-features.md) to see test-first approach in work for top-down approach.

For the documention of code-first approach, I would like to see a similar content/narrative:

 1. Pick one (and only one!) endpoint you are going to implement.
 2. Start with an acceptance test describing endpoint's behaviour (see [Write an acceptance test](https://github.com/strongloop/loopback.io/blob/7ccfb55463b36f9177a7960926f933ae47f00986/pages/en/lb4/Implementing-features.md#write-an-acceptance-test)). This acceptance test is going to server as an indicator telling you when you are done with this endpoint.
 3. Write a unit-test for the new controller method you are going to implement (see [Write a unit-test for the new controller method](https://github.com/strongloop/loopback.io/blob/7ccfb55463b36f9177a7960926f933ae47f00986/pages/en/lb4/Implementing-features.md#write-a-unit-test-for-the-new-controller-method), this section contains an example of the actual controller implementation at the bottom.)
 4. Write the first implementation of the controller method to make the test pass. This is time to put code-first approach in work, e.g. use REST decorators like `@get` and `@param`.
 5. Write an integration test to verify that the controller is using data from the database (see [Write an integration test for the new controller method](https://github.com/strongloop/loopback.io/blob/7ccfb55463b36f9177a7960926f933ae47f00986/pages/en/lb4/Implementing-features.md#write-an-integration-test-for-the-new-controller-method)
 6. Now comes the best time for defining models & repositories - see [Define Product model, repository, and data source](https://github.com/strongloop/loopback.io/blob/7ccfb55463b36f9177a7960926f933ae47f00986/pages/en/lb4/Implementing-features.md#define-product-model-repository-and-data-source). Because you are describing bottom-up approach, I think it's acceptable to assume that our data/repository/model layer is already set up and/or was configured before we started writing controllers for the REST API. This configuration should still be implemented in test-first way though!
 7. [Update test helpers and the controller use real model and repository](https://github.com/strongloop/loopback.io/blob/7ccfb55463b36f9177a7960926f933ae47f00986/pages/en/lb4/Implementing-features.md#update-test-helpers-and-the-controller-use-real-model-and-repository)
 8. Run the tests again and fix the issues - see [Run tests](https://github.com/strongloop/loopback.io/blob/7ccfb55463b36f9177a7960926f933ae47f00986/pages/en/lb4/Implementing-features.md#run-tests)
 9. [Decouple controller from Repository](https://github.com/strongloop/loopback.io/blob/7ccfb55463b36f9177a7960926f933ae47f00986/pages/en/lb4/Implementing-features.md#decouple-controller-from-repository)
 10. [Handle 'product not found' error](https://github.com/strongloop/loopback.io/blob/7ccfb55463b36f9177a7960926f933ae47f00986/pages/en/lb4/Implementing-features.md#handle-product-not-found-error)

Basically I am asking you to create a guide that's similar to what we have for bottom-down approach, but with specific details changed to accommodate the bottom-up (code-first) approach.

If there are sections that can or should be reused by both guides, then it may be worth extracting them into shared content. (I don't know how easy it is, or whether it's feasible at all.)

For example, the last step in the api-first guide, [Implement a custom sequence](https://github.com/strongloop/loopback.io/blob/7ccfb55463b36f9177a7960926f933ae47f00986/pages/en/lb4/Implementing-features.md#implement-a-custom-sequence), should be pretty easy to extract into a standalone document that both code-first and api-first guide can refer to.

I am not sure what's the exact scope of your task (the user story you are working on). If a guide as I described above is beyond the scope, then I can live with the shorter version you proposed here and will open a new issue to bring the code-first guide up to the level of the api-first guide. Please let me know!

**Comment by shimks** [comment]
Thanks for a very detailed narrative @bajtos ! Considering that @kjdelisle is working on the tutorial story (https://github.com/strongloop/loopback-next/issues/729), I personally thought the story I'm working on (https://github.com/strongloop/loopback-next/issues/735) would serve to merely accompany the `Defining and validating the API` page with the bottom-up approach. It seems like the narrative you proposed requires the docs I'm writing to be essentially an extension piece of `example-getting-started` that shows how the tutorial is written the test-first approach.

Right now I'm thinking that I rewrite this section as an __extension piece__ of the tutorial, continuing off from https://github.com/strongloop/loopback-next/blob/master/packages/example-getting-started/docs/6-repository.md. From there I'm thinking of following your proposed content structure to finish off `example-getting-started` application with complete tests to go with it. Does this sound good to you?

File: pages/en/lb4/Best-practices-with-Loopback-4.md (hunk 21 above)
Comment (shimks) [reply pages/en/lb4/Best-practices-with-Loopback-4.md]: Back when I worked on breaking up the `Thinking in LoopBack` page, the rough structure of the page went from:
1. Define your API
2. Test your __API__
3. Summary of TTD
4. Writing your code using TTD

To me testing the app is different from testing the API (spec), and it makes sense to me that when you're using the build-your-API-spec-first approach you would go about testing the API (in the form of openapi spec) first before you test anything that has to do with the application.

File: _data/sidebars/lb4_sidebar.yml (hunk 18 above)
Comment (shimks) [reply _data/sidebars/lb4_sidebar.yml:114]: +1. As a helpful comment though, ~~never~~ try not to use `'s` to make something plural :p, even for acronyms (unless it ends with an S, then I have no clue lol).

**Comment by bajtos** [comment]
>  merely accompany the Defining and validating the API page with the bottom-up approach.

I see. I quickly checked the content in [Defining and validating the API](https://github.com/strongloop/loopback.io/blob/7ccfb55463b36f9177a7960926f933ae47f00986/pages/en/lb4/Defining-and-validating-the-API.md), I agree your current content is closer to that page than my long proposal, which makes more sense in the context of your task.

So maybe it's enough if you add notes to your existing content to tell people that the examples in your content are just showing how to define API and pointing them to "Incrementally implement features" for better examples showing the implementation process including test coverage?

> Right now I'm thinking that I rewrite this section as an extension piece of the tutorial, continuing off from https://github.com/strongloop/loopback-next/blob/master/packages/example-getting-started/docs/6-repository.md. From there I'm thinking of following your proposed content structure to finish off example-getting-started application with complete tests to go with it. Does this sound good to you?

Sounds reasonable to me. I think it may be best to split the work into two (or more) pull requests, the first one covering "I rewrite this section as an extension piece of the tutorial", so that we can land something sooner than later?

**Comment by shimks** [comment]
> I think it may be best to split the work into two (or more) pull requests, the first one covering "I rewrite this section as an extension piece of the tutorial", so that we can land something sooner than later?

Could you clarify how I could go about breaking up the work into multiple pull requests? Are you thinking that the first PR to be less detailed than what it should be? Or do you mean I should break up the actual contents in chunks that resembles something like this ?:
`acceptance test -> unit test -> write logic` -> `integration test -> update tests/code -> tahdah you're done!!!`

**Comment by bajtos** [comment]
> Could you clarify how I could go about breaking up the work into multiple pull requests? Are you thinking that the first PR to be less detailed than what it should be? 

As I was reading your comment, it occurred to me that your plan seems to have two distinctive parts/tasks:

> Right now I'm thinking that I rewrite this section as an extension piece of the tutorial, continuing off from https://github.com/strongloop/loopback-next/blob/master/packages/example-getting-started/docs/6-repository.md. From there I'm thinking of following your proposed content structure to finish off example-getting-started application with complete tests to go with it. Does this sound good to you?

I was thinking:

 - pull request 1:  _rewrite this section as an extension piece of the tutorial, continuing off from https://github.com/strongloop/loopback-next/blob/master/packages/example-getting-started/docs/6-repository.md_
 - pull request 2 (3, 4, etc.): _From there I'm thinking of following your proposed content structure to finish off example-getting-started application with complete tests to go with it._

Just an idea to consider. Please organize your work and pull requests in a way that works best for you!

**Comment by bajtos** [comment]
@shimks I think we should find a smaller scope for your work, considering the description in https://github.com/strongloop/loopback-next/issues/735:

> Add a new page at the same level, something like "Define your API bottom-up way". Explain basic concepts (controller decorators like @post, @param etc.) and point people to example-getting-started as the tutorial.

The outline described in https://github.com/strongloop/loopback.io/pull/624#issuecomment-369612424 may require more effort that we have planned for this task. Could you please talk to @kjdelisle to find a good balance between the ideal I described and the MVP/Walking Skeleton scope.

File: pages/en/lb4/Best-practices-with-Loopback-4.md [hunk 22]
```diff
@@ -8,12 +8,21 @@ permalink: /doc/en/lb4/Best-practices-with-Loopback-4.html
...
-4. [**Implementing features**](./Implementing-features.html): This section demonstrates how the tests for each feature of your application should be written, and how to write the logic to make these tests pass. In the example, the tests for the controller, model, repository, data source, and sequence are written and then implemented.
-5. [**Preparing the API for consumption**](./Preparing-the-API-for-consumption.html): This section shows how the endpoints can be physically tested using the Swagger UI.
+1. **Defining the API**: There are two possible approaches to take in this section
+   - [**Defining the API using code-first approach**](./Defining-the-API-using-code-first-approach.html): This section guides you through setting up a skeleton of your application so that its full OpenAPI specification can be automatically generated.
+   - [**Defining the API using API-first approach**](./Defining-the-API-using-API-first-approach.html): This section guides you through constructing your API first before any internal logic is added. __*Not fully supported*__
```
Comment (kjdelisle) [inline pages/en/lb4/Best-practices-with-Loopback-4.md]: I'd rephrase this bit to say "Defining your app using the API-first approach" or something similar.
"Defining the API using API-first approach" sounds a bit like saying "Making your app by using the app-first approach" :P

Comment (kjdelisle) [inline pages/en/lb4/Best-practices-with-Loopback-4.md]: I'd rephrase this bit to say "Defining your app using the API-first approach" or something similar.
"Defining the API using API-first approach" sounds a bit like saying "Making your app by using the app-first approach" :P

File: pages/en/lb4/Defining-the-API-using-API-first-approach.md [hunk 23]
```diff
@@ -13,7 +13,7 @@ page are outdated and may not work out of the box. They will be revisited after
 our MVP release.
 "%}
 
-## Define the API
+## Define the API from top to bottom (API-first)
```
Comment (kjdelisle) [inline pages/en/lb4/Defining-the-API-using-API-first-approach.md]: Define the *app* from top to bottom

**Review by kjdelisle — CHANGES_REQUESTED** [review]
In all cases, we're defining an application, not an API. The API is a part of the application (since it's what consumers will use), but the phrasing on a lot of the top-down methods is wobbly if you refer to the whole app as an API.

File: _data/sidebars/lb4_sidebar.yml [hunk 24]
```diff
@@ -111,13 +111,18 @@ children:
...
-  - title: 'Defining and validating the API'
-    url: Defining-and-validating-the-API.html
+  - title: 'Defining the API using code-first approach'
+    url: Defining-the-API-using-code-first-approach.html
     output: 'web, pdf'
 
-  - title: 'Testing the API'
-    url: Testing-the-API.html
+  - title: 'Defining the app using API-first approach'
+    url: Defining-the-app-using-API-first-approach.html
```
Comment (raymondfeng) [inline _data/sidebars/lb4_sidebar.yml]: What about `Design First vs. Code First`? See https://swaggerhub.com/blog/api-design/design-first-or-code-first-api-development/

**Review by jannyHou — APPROVED** [review]
LGTM 👍

File: pages/en/lb4/Defining-the-API-using-code-first-approach.md [hunk 25]
```diff
@@ -0,0 +1,186 @@
...
+Once your routes have been decorated, your application is ready to serve
+its API. When an instance of `RestServer` is run, an OpenAPI specification
+representing your application's API is built. The spec is generated
+entirely from the decorated elements' metadata, which in turn provides
+routing logic for your API when your application is running.
+
+### Reviewing your API specification
+
+To review your complete API specification, run your application with the
+decorated controllers registered. Once it is running, visit `/swagger.json`
```
Comment (b-admike) [inline pages/en/lb4/Defining-the-API-using-code-first-approach.md]: `/openapi.json`

**Review by bschrammIBM — APPROVED** [review]
I will need to copyedit this doc after it is posted due to time constraints with Velox.
